plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ElasticsearchMetadata.java

```diff
@@ -108,6 +120,84 @@ public ElasticsearchMetadata(TypeManager typeManager, ElasticsearchClient client
         this.schemaName = config.getDefaultSchema();
     }
 
+    @Override
+    public Optional<AggregationApplicationResult<ConnectorTableHandle>> applyAggregation(
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            List<AggregateFunction> aggregates,
+            Map<String, ColumnHandle> assignments,
+            List<List<ColumnHandle>> groupingSets)
+    {
+        if (!session.getProperty(AGGREGATION_PUSHDOWN_ENABLED, Boolean.class)) {
+            return Optional.empty();
+        }
+        ElasticsearchTableHandle handle = (ElasticsearchTableHandle) table;
+        if (isPassthroughQuery(handle)) {
+            // aggregation pushdown currently not supported passthrough query
+            return Optional.empty();
+        }
+        Version version = client.getVersion();
+        // in order to support null bucket, we need missing_bucket in composite aggregation which was introduced in 6.4.0
+        if (version == null || version.before(ElasticsearchClient.MINIMUM_VERSION_REQUIRE_FOR_AGG_PUSHDOWN)) {
+            return Optional.empty();
+        }
+        // Global aggregation is represented by [[]]
+        verify(!groupingSets.isEmpty(), "No grouping sets provided");
+        if (handle.getTermAggregations() != null && !handle.getTermAggregations().isEmpty()) {
+            // applyAggregation may be called multiple times target on the same ElasticsearchTableHandle
+            // for example
+            // SELECT sum(DISTINCT regionkey) FROM nation
+            // will be split into two logic phrase:
+            // 1. table a -> select regionkey from nation group by regionkey
+            // 2. select sum(regionkey) from a
+            // so the first call of applyAggregation will only contains groupby set
+            // and the second call will contains aggregates(sum)
+            // we skip the second one, but the first group by will be still applied
+            return Optional.empty();
```

Comment:
This comment is incorrect. There will be a single call for the whole GROUP BY phase. You might see multiple calls to `applyAggregation` if an aggregation is done over the results of another aggregation. E.g.,

```
SELECT max(x)
FROM (
    SELECT k, sum(v) AS x
    FROM t
    GROUP BY k
)
```